File: JavaScript/mict_ai_ethics.js
// javascript/mict-ai-ethics.js

/**
 * Calculates the disparate impact ratio for a binary classification model.
 * @param {number[]} predictions - Array of predicted labels (0 or 1).
 * @param {number[]} labels - Array of true labels (0 or 1).
 * @param {number[]} sensitiveAttribute - Array indicating group membership (0 or 1).
 * @returns {number|null} The disparate impact ratio, or null if undefined.
 */
function calculateDisparateImpact(predictions, labels, sensitiveAttribute) {
    if (predictions.length !== labels.length || predictions.length !== sensitiveAttribute.length) {
        return null; // Or throw an error
    }

    const group0PredPositive = predictions.filter((p, i) => sensitiveAttribute[i] === 0 && p === 1).length;
    const group0Total = sensitiveAttribute.filter(a => a === 0).length;
    const group1PredPositive = predictions.filter((p, i) => sensitiveAttribute[i] === 1 && p === 1).length;
    const group1Total = sensitiveAttribute.filter(a => a === 1).length;

    if (group0Total === 0 || group1Total === 0) {
        return null; // Avoid division by zero
    }

    const group0Rate = group0PredPositive / group0Total;
    const group1Rate = group1PredPositive / group1Total;

    if (group0Rate === 0) {
        return null; // Avoid division by zero
    }

    return group1Rate / group0Rate;
}

/**
 * Simple hallucination detection based on string matching against a knowledge base.
 * @param {string} text - The generated text.
 * @param {string[]} knowledgeBase - An array of known facts.
 * @returns {string[]} An array of potentially hallucinated statements.
 */
function detectHallucinations(text, knowledgeBase) {
  const sentences = text.split(/(?<=[.!?])\s+/); // Split into sentences.
  const hallucinations = [];

  for (const sentence of sentences) {
    let found = false;
    for (const fact of knowledgeBase) {
      if (sentence.toLowerCase().includes(fact.toLowerCase())) {
        found = true;
        break;
      }
    }
    if (!found) {
      hallucinations.push(sentence);
    }
  }
  return hallucinations;
}

// --- Example Usage (within a MICT Cycle) ---
// This would typically be in an example file, NOT in mict-ai-ethics.js itself

// import { calculateDisparateImpact, detectHallucinations } from './mict-ai-ethics.js';
// import { createMICTEngine } from './mict-framework.js'; // Assuming you have a separate file

// ... (MICT cycle setup, stage functions, etc.) ...

export { calculateDisparateImpact, detectHallucinations };

File: Python/mict_ai_ethics.py
# python/mict_ai_ethics.py

from typing import List, Optional, Dict, Any

def calculate_disparate_impact(predictions: List[int], labels: List[int], sensitive_attribute: List[int]) -> Optional[float]:
    """
    Calculates the disparate impact ratio for a binary classification model.

    Args:
        predictions: List of predicted labels (0 or 1).
        labels: List of true labels (0 or 1).
        sensitive_attribute: List indicating group membership (0 or 1).

    Returns:
        The disparate impact ratio, or None if undefined (e.g., division by zero).
    """
    if len(predictions) != len(labels) or len(predictions) != len(sensitive_attribute):
        return None  # Or raise an exception

    group0_pred_positive = sum(1 for p, s in zip(predictions, sensitive_attribute) if s == 0 and p == 1)
    group0_total = sensitive_attribute.count(0)
    group1_pred_positive = sum(1 for p, s in zip(predictions, sensitive_attribute) if s == 1 and p == 1)
    group1_total = sensitive_attribute.count(1)

    if group0_total == 0 or group1_total == 0:
        return None  # Avoid division by zero

    group0_rate = group0_pred_positive / group0_total
    group1_rate = group1_pred_positive / group1_total

    if group0_rate == 0:
        return None  # Avoid division by zero

    return group1_rate / group0_rate


def detect_hallucinations(text: str, knowledge_base: List[str]) -> List[str]:
    """
    Simple hallucination detection based on string matching against a knowledge base.

    Args:
        text: The generated text.
        knowledge_base: A list of known facts.

    Returns:
        A list of potentially hallucinated statements.
    """
    sentences = text.split(".")  # VERY simplistic sentence splitting
    hallucinations = []

    for sentence in sentences:
        sentence = sentence.strip()
        if not sentence:  # Skip empty sentences
            continue
        found = False
        for fact in knowledge_base:
            if fact.lower() in sentence.lower():
                found = True
                break
        if not found:
            hallucinations.append(sentence)

    return hallucinations
